# Assessment

### 🧠 Task:

1. **Data Cleaning & Exploration**

   * Perform basic EDA (Exploratory Data Analysis)
   * Identify any data quality issues or assumptions

2. **Merchant Performance Metrics**

   * Define and calculate relevant KPIs for merchant performance (e.g., number of active members, avg spend per member, retention rate)
   * Identify top and bottom merchants using your metrics

3. **Insights & Recommendations**

   * Summarize key business insights
   * Recommend what actions should be taken for underperforming merchants

4. **Dashboard/Visualization**

   * Build a simple dashboard to communicate your findings (in Tableau Public or screenshots if using other tools)
   * Include filters or interactive components if applicable

# Library

In [1]:
import pandas as pd
import os
import matplotlib.pyplot as plt

In [2]:
os.chdir(r'C:\Users\User\OneDrive\文档\Work\data-takehome-test_bi-main')

In [3]:
os.listdir()

csv_files = [f for f in os.listdir() if f.endswith('.csv')]
# Load all CSVs into dataframes and store them in a dictionary
dfs = {os.path.splitext(f)[0]: pd.read_csv(f) for f in csv_files}

df_members = dfs['members']
df_merchants = dfs['merchants']
df_transactions = dfs['transactions']
df_engagements = dfs['engagement']

In [4]:
# 1. Data Cleaning & Exploration
# - Check for missing values, duplicates, and data types
display(df_members.info())
display(df_merchants.info())
display(df_transactions.info())
display(df_engagements.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   member_id    500 non-null    object
 1   signup_date  500 non-null    object
 2   tier         500 non-null    object
 3   age_group    500 non-null    object
 4   state        500 non-null    object
dtypes: object(5)
memory usage: 19.7+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   merchant_id    20 non-null     object
 1   merchant_name  20 non-null     object
 2   category       20 non-null     object
 3   location       20 non-null     object
 4   join_date      20 non-null     object
dtypes: object(5)
memory usage: 932.0+ bytes


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   transaction_id  3000 non-null   object 
 1   member_id       3000 non-null   object 
 2   merchant_id     3000 non-null   object 
 3   spend_amount    3000 non-null   float64
 4   timestamp       3000 non-null   object 
dtypes: float64(1), object(4)
memory usage: 117.3+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   member_id       500 non-null    object
 1   monthly_logins  500 non-null    int64 
 2   redemptions     500 non-null    int64 
 3   app_opens       500 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 15.8+ KB


None

In [7]:
display("df_members head:")
display(df_members.head())

display("df_merchants head:")
display(df_merchants.head())

display("df_transactions head:")
display(df_transactions.head())

display("df_engagements head:")
display(df_engagements.head())

'df_members head:'

member_id signup_date      tier age_group         state
0     M0001  2022-04-13      Gold       55+         Sabah
1     M0002  2022-12-15    Silver     18-24        Penang
2     M0003  2022-09-28    Silver       55+  Kuala Lumpur
3     M0004  2022-04-17  Platinum       55+         Johor
4     M0005  2022-03-13    Silver     35-44        Penang

'df_merchants head:'

merchant_id merchant_name   category location   join_date
0      MER001    Merchant_1  eCommerce   Penang  2022-04-11
1      MER002    Merchant_2  eCommerce       PJ  2021-11-07
2      MER003    Merchant_3  eCommerce       KK  2021-09-19
3      MER004    Merchant_4  eCommerce   Penang  2022-02-23
4      MER005    Merchant_5     Retail       KL  2022-02-10

'df_transactions head:'

transaction_id member_id merchant_id  spend_amount   timestamp
0         T00001     M0076      MER010          9.76  2023-06-07
1         T00002     M0385      MER010          5.81  2023-05-14
2         T00003     M0439      MER020         41.85  2023-03-08
3         T00004     M0278      MER009        131.79  2023-12-02
4         T00005     M0040      MER017          2.82  2023-01-07

'df_engagements head:'

member_id  monthly_logins  redemptions  app_opens
0     M0001               4            1         10
1     M0002               4            1          7
2     M0003               6            2         11
3     M0004               3            1         16
4     M0005               3            1          7

In [ ]:
# Analyze member distribution by state and age group

# Members by state
state_counts = df_members['state'].value_counts()
print("Member count by state:")
display(state_counts)

# Members by age group
age_group_counts = df_members['age_group'].value_counts()
print("\nMember count by age group:")
display(age_group_counts)

# Signup analysis: members by signup month
df_members['signup_date'] = pd.to_datetime(df_members['signup_date'])
signup_month = df_members['signup_date'].dt.to_period('M').value_counts().sort_index()
print("\nMember signups by month:")
display(signup_month)

# Cross-tabulation of state and age group
state_age_group = pd.crosstab(df_members['state'], df_members['age_group'])
print("\nMembers by state and age group:")
display(state_age_group)

Member count by state:


state
Johor           111
Kuala Lumpur    108
Sabah           100
Selangor        100
Penang           81
Name: count, dtype: int64


Member count by age group:


age_group
55+      107
35-44    106
25-34    106
18-24     98
45-54     83
Name: count, dtype: int64


Member signups by month:


signup_date
2022-01    31
2022-02    47
2022-03    23
2022-04    49
2022-05    51
2022-06    44
2022-07    37
2022-08    47
2022-09    45
2022-10    43
2022-11    38
2022-12    45
Freq: M, Name: count, dtype: int64


Members by state and age group:


age_group     18-24  25-34  35-44  45-54  55+
state                                        
Johor            25     28     23     15   20
Kuala Lumpur     26     17     24     13   28
Penang           20     13     18     14   16
Sabah             9     26     22     25   18
Selangor         18     22     19     16   25

### Analysis of Member Distribution Patterns

#### 1. State Distribution
Johor and Kuala Lumpur have the highest number of members (111 and 108), followed by Sabah and Selangor (100 each), and Penang (81).
This suggests that Johor and KL are key markets with higher member engagement or acquisition.

#### 2. Age Group Distribution
The largest age group is 55+ (107 members), followed closely by 35-44 and 25-34 (106 each), then 18-24 (98), and 45-54 (83).
This indicates a relatively older member base, with a strong presence in the 55+ segment.

#### 3. State vs Age Group (state_age_group)
- KL & Selangor have a significant number of members in the 55+ age group.
- Sabah has a more balanced distribution across age groups, with higher numbers in 25-34, 35-44, and 45-54.
- Johor and KL have strong representation in the 18-24 and 55+ groups.
- Penang has the lowest member count overall, with a relatively even spread across age groups.

#### 4. Signup Trends (signup_month)
There are spikes in member signups in April, May, August, and December 2022.
This could be due to marketing campaigns, seasonal promotions, or other external factors.

##### Summary of Patterns:
- Johor and KL are the largest states by member count.
- The 55+ age group is the largest, especially in KL and Selangor.
- Sabah has a more even age distribution.
- Signup activity varies by month, with some clear peaks.

##### Recommendations for further analysis:
- Investigate what drives higher signups in certain months.
- Explore engagement and transaction patterns by age group and state.
- Consider targeted campaigns for underrepresented states (e.g., Penang) or age groups (e.g., 45-54).

In [ ]:
# Explore the remaining dataframes: df_merchants, df_transactions, df_engagements

# Merchants: unique categories, locations, join date range
print("Merchant categories:")
display(df_merchants['category'].value_counts())
print("\nMerchant locations:")
display(df_merchants['location'].value_counts())
print("\nMerchant join date range:")
display(df_merchants['join_date'].min(), df_merchants['join_date'].max())

# Transactions: spend amount distribution, transaction date range, unique merchants/members
print("\nTransaction spend amount stats:")
display(df_transactions['spend_amount'].describe())
print("\nTransaction date range:")
display(df_transactions['timestamp'].min(), df_transactions['timestamp'].max())
print("\nUnique merchants in transactions:", df_transactions['merchant_id'].nunique())
print("Unique members in transactions:", df_transactions['member_id'].nunique())

# Engagements: summary statistics
print("\nEngagement metrics summary:")
display(df_engagements.describe())
print("\nEngagements - members with zero redemptions:")
display((df_engagements['redemptions'] == 0).sum())

Merchant categories:


category
eCommerce    9
Retail       5
F&B          5
Travel       1
Name: count, dtype: int64


Merchant locations:


location
PJ        5
KL        5
Penang    4
KK        3
JB        3
Name: count, dtype: int64


Merchant join date range:


'2021-01-12'

'2022-11-09'


Transaction spend amount stats:


count    3000.00000
mean       50.07977
std        50.56371
min         0.01000
25%        15.09000
50%        34.53500
75%        68.75750
max       412.99000
Name: spend_amount, dtype: float64


Transaction date range:


'2023-01-01'

'2023-12-31'


Unique merchants in transactions: 20
Unique members in transactions: 498

Engagement metrics summary:


monthly_logins  redemptions   app_opens
count      500.000000   500.000000  500.000000
mean         4.006000     1.586000   10.042000
std          1.978331     1.047281    3.017025
min          0.000000     0.000000    3.000000
25%          3.000000     1.000000    8.000000
50%          4.000000     1.000000   10.000000
75%          5.000000     2.000000   12.000000
max         11.000000     5.000000   21.000000


Engagements - members with zero redemptions:


np.int64(71)

In [17]:
# 2. Merchant Performance Metrics
# - Number of active members per merchant
active_members = df_transactions.groupby('merchant_id')['member_id'].nunique()

# - Average spend per member per merchant
avg_spend_per_member = df_transactions.groupby('merchant_id').apply(
    lambda x: x['spend_amount'].sum() / x['member_id'].nunique()
)

# - Retention rate (e.g., members with >1 transaction at the same merchant)
retained = df_transactions.groupby(['merchant_id', 'member_id']).size().reset_index(name='tx_count')
retention = retained[retained['tx_count'] > 1].groupby('merchant_id').size() / active_members

# - Combine metrics into a summary DataFrame
merchant_metrics = pd.DataFrame({
    'active_members': active_members,
    'avg_spend_per_member': avg_spend_per_member,
    'retention_rate': retention
}).fillna(0)

# - Identify top and bottom merchants
top_merchants = merchant_metrics.sort_values('avg_spend_per_member', ascending=False).head()
bottom_merchants = merchant_metrics.sort_values('avg_spend_per_member').head()

C:\Users\User\AppData\Local\Temp\ipykernel_25924\3578433279.py:6: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  avg_spend_per_member = df_transactions.groupby('merchant_id').apply(


In [18]:
top_merchants

active_members  avg_spend_per_member  retention_rate
merchant_id                                                      
MER014                  120             64.386417        0.133333
MER018                  137             63.879562        0.189781
MER020                  122             62.311721        0.155738
MER010                  140             61.153000        0.121429
MER012                  123             60.459187        0.146341

In [19]:
bottom_merchants

active_members  avg_spend_per_member  retention_rate
merchant_id                                                      
MER019                  130             50.605769        0.084615
MER009                  135             51.028815        0.118519
MER017                  143             52.404615        0.125874
MER001                  149             52.775302        0.114094
MER015                  125             55.014320        0.120000

## Summary of Key Findings

- **Member Distribution:** Johor (111) and Kuala Lumpur (108) have the highest member counts, followed by Sabah and Selangor (100 each), and Penang (81).
- **Age Groups:** The largest segment is 55+ (107), followed by 35-44 (106), 25-34 (106), 18-24 (98), and 45-54 (83). The member base skews older.
- **State vs Age:** KL and Selangor have many 55+ members; Sabah’s distribution is more balanced across age groups.
- **Signup Trends:** Member signups peaked in April, May, August, and December 2022, likely due to campaigns or seasonality.
- **Merchants:** 20 merchants across various categories and locations, with join dates from early 2021 to late 2022.
- **Transactions:** 3,000 transactions, with a wide range of spend amounts and active participation from both members and merchants.
- **Engagement:** Most members have some engagement, but a notable number have zero redemptions.

**Merchant Performance Analysis:**
- **Active Members per Merchant:** There is significant variation in the number of unique members transacting with each merchant.
- **Average Spend per Member:** Top merchants have much higher average spend per member, indicating strong customer value, while bottom merchants lag behind.
- **Retention Rate:** Some merchants have higher retention rates (members making repeat purchases), while others struggle to retain customers.
- **Top Merchants:** Identified by highest average spend per member and strong retention.
- **Bottom Merchants:** Identified by lowest average spend per member and low retention.

**Insights:**
- The platform is strongest in Johor and KL, especially among older members.
- Some states (e.g., Penang) and age groups (e.g., 45-54) are underrepresented, suggesting opportunities for targeted growth.
- Engagement and transaction activity is healthy, but some members are not redeeming, indicating potential for improved engagement strategies.
- Merchant performance is uneven; a few merchants drive most value, while others underperform in both spend and retention.

**Recommendations:**
- Target marketing to underrepresented states and age groups.
- Investigate what drove signup spikes for future campaigns.
- Encourage engagement and redemptions among less active members.
- Support underperforming merchants with tailored strategies (e.g., promotions, loyalty incentives) to boost member activity and retention.
- Analyze top merchant practices and replicate successful tactics across the merchant base.